src/reactor/poll_evented2.rs

```diff
@@ -0,0 +1,419 @@
+//! acquisition of a token, and tracking of the readiness state on the
+//! underlying I/O primitive.
+
+#![allow(warnings)]
+
+use reactor::Handle;
+use reactor::registration::Registration;
+
+use futures::{task, Async, Poll};
+use mio;
+use mio::event::Evented;
+use tokio_io::{AsyncRead, AsyncWrite};
+
+use std::fmt;
+use std::io::{self, Read, Write};
+use std::sync::atomic::AtomicUsize;
+use std::sync::atomic::Ordering::Relaxed;
+
+/// A concrete implementation of a stream of readiness notifications for I/O
+/// objects that originates from an event loop.
+///
+/// Created by the `PollEvented::new` method, each `PollEvented` is
+/// associated with a specific event loop and source of events that will be
+/// registered with an event loop.
+///
+/// An instance of `PollEvented` is essentially the bridge between the `mio`
+/// world and the `tokio-core` world, providing abstractions to receive
+/// notifications about changes to an object's `mio::Ready` state.
+///
+/// Each readiness stream has a number of methods to test whether the underlying
+/// object is readable or writable. Once the methods return that an object is
+/// readable/writable, then it will continue to do so until the `need_read` or
+/// `need_write` methods are called.
+///
+/// That is, this object is typically wrapped in another form of I/O object.
+/// It's the responsibility of the wrapper to inform the readiness stream when a
+/// "would block" I/O event is seen. The readiness stream will then take care of
+/// any scheduling necessary to get notified when the event is ready again.
+///
+/// You can find more information about creating a custom I/O object [online].
+///
+/// [online]: https://tokio.rs/docs/going-deeper-tokio/core-low-level/#custom-io
+///
+/// ## Readiness to read/write
+///
+/// A `PollEvented` allows listening and waiting for an arbitrary `mio::Ready`
+/// instance, including the platform-specific contents of `mio::Ready`. At most
+/// two future tasks, however, can be waiting on a `PollEvented`. The
+/// `need_read` and `need_write` methods can block two separate tasks, one on
+/// reading and one on writing. Not all I/O events correspond to read/write,
+/// however!
+///
+/// To account for this a `PollEvented` gets a little interesting when working
+/// with an arbitrary instance of `mio::Ready` that may not map precisely to
+/// "write" and "read" tasks. Currently it is defined that instances of
+/// `mio::Ready` that do *not* return true from `is_writable` are all notified
+/// through `need_read`, or the read task.
+///
+/// In other words, `poll_ready` with the `mio::UnixReady::hup` event will block
+/// the read task of this `PollEvented` if the `hup` event isn't available.
+/// Essentially a good rule of thumb is that if you're using the `poll_ready`
+/// method you want to also use `need_read` to signal blocking and you should
+/// otherwise probably avoid using two tasks on the same `PollEvented`.
+pub struct PollEvented<E> {
```

Comment:
The existing `PollEvented` type is not compatible with lazy binding due to how the trait bounds are setup. Because of this, a new type must be introduced and the existing `PollEvented` deprecated.

This also reverts some aspects of `PollEvented` back to what they were originally in `tokio-core`. Adding lazy binding reduced some of the benefits of limiting fns to `&mut self`.